### ISSUE TITLE
How to save Keras model with custom layer that have dictionary variable?

### ISSUE THREAD

**Issue by saylideshmukh** [keras-team/keras#16276]
Please go to TF Forum for help and support:

https://discuss.tensorflow.org/tag/keras

If you open a GitHub issue, here is our policy:

It must be a bug, a feature request, or a significant problem with the documentation (for small docs fixes please send a PR instead).
The form below must be filled out.

**Here's why we have that policy:**.

Keras developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):2.3
- Python version:3.6
- Bazel version (if compiling from source):
- GPU model and memory:
- Exact command to reproduce:

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with:
python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"

**Describe the problem**.
I have a custom layer , which has a dictionary variable called vmap. I implement the get_config() and from_config() method in the custom layer class and try to save the model using : 
model.save('model.h5') 
and loading the model using : 
new_model = tf.keras.models.load_model('model.h5',custom_objects = {"valuemaplayer": valuemaplayer})
In the config method if i  do "vmap": self._vmap and try to save the model it is saved and loaded successfully. However, i need to save the dictionary vmap with its content so i do "vmap": self._vmap.items() in config method and try to retrain and save that retrained model but i get an error stating  Type error: keys must be str,int,float,bool or None , not Tuple.
Describe the problem clearly here. Be sure to convey here why it's a bug in Keras or why the requested feature is needed.
This is my code for custom layer : 
class valuemaplayer(keras.layers.Layer):
    def __init__(self,vmap ={},compress = False, **kwargs):
        kwargs["dynamic"] = True
        super(valuemaplayer,self).__init__(**kwargs)
        self._vmap = vmap
        self._data = []
        self._compression = compress
        self._output_shape = None

    def build(self, input_shape):
        pass

    def enable_compression(self):
        value = list(self.get_values())
        vmap = self._vmap
        cnt = 0
        # FIXME right now only 2D input data is supported
        for v0 in value:
            for v1 in v0:
                for v2 in v1:
                    for v3 in v2:
                        v = tuple(v3)
                        if v not in vmap:
                            vmap[v]=cnt
                            cnt+=1
        self._compression = True
        print(vmap)

    @tf.function
    def do_mapping(self,pixel):
        enumerated_value = self._vmap.get(pixel)
            # print(enumerated_value)
            # print(pixel)
        return enumerated_value

    def call(self, inputs, training=True):#use eager execution or decorate with @tf.function
        if self._compression:
            elem = []
            for b in inputs:
                for h in b:
                    for w in h:
                        x = self.do_mapping(pixel=tuple(w.numpy()))
                        elem.append(tf.convert_to_tensor(x, dtype='float32'))
            return tf.cast(tf.reshape(elem, self._output_shape), dtype='float32')

            # TODO check if channel axis gets mapped by tf.map_fn
         # else compression is disabled
         # in case we're training, we do not want to observe values
        if not training:
            self._data.append(inputs)
        return inputs

        # get values of the output of value map layer
    def get_values(self):
        for d in self._data:
            try:
                d = d.numpy()
            except AttributeError:
                continue
            yield d

    def get_config(self):
        config = super(valuemaplayer, self).get_config()
        config.update({"vmap": self._vmap.items(),
                       "compress": self._compression})
        return config

    @classmethod
    def from_config(cls, config):
        return cls(**config)

    def compute_output_shape(self, input_shape):
        print("input shape of value map layer:", input_shape)
        self._output_shape=input_shape
        b, h, w, _ = input_shape
        if self._compression:
            # TODO did I set the channel axis? and does this work?
            print("channel axis is set")
            self._output_shape = tf.TensorShape((b, h, w, 1))
        return self._output_shape

#saving the model : 
model.save('model.h5')
#load model :
# new_model = tf.keras.models.load_model('model.h5',custom_objects = {"valuemaplayer": valuemaplayer})
**Describe the current behavior**.
Type error: keys must be str,int,float,bool or None , not Tuple.

**Describe the expected behavior**.
Model saved successfully with the custom layer and its dictionary variable vmap.
**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Source code / logs**.

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by sushreebarsa** [comment]
@saylideshmukh 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by saylideshmukh** [comment]
Hello @sushreebarsa 
This is my code for custom layer :
class valuemaplayer(keras.layers.Layer):
def init(self,vmap ={},compress = False, **kwargs):
kwargs["dynamic"] = True
super(valuemaplayer,self).init(**kwargs)
self._vmap = vmap
self._data = []
self._compression = compress
self._output_shape = None

def build(self, input_shape):
    pass

def enable_compression(self):
    value = list(self.get_values())
    vmap = self._vmap
    cnt = 0
    # FIXME right now only 2D input data is supported
    for v0 in value:
        for v1 in v0:
            for v2 in v1:
                for v3 in v2:
                    v = tuple(v3)
                    if v not in vmap:
                        vmap[v]=cnt
                        cnt+=1
    self._compression = True
    print(vmap)


@tf.function
def do_mapping(self,pixel):
    enumerated_value = self._vmap.get(pixel)
    return enumerated_value

def call(self, inputs, training=True):#use eager execution or decorate with @tf.function
    if self._compression:
        elem = []
        for b in inputs:
            for h in b:
                for w in h:
                    x = self.do_mapping(pixel=tuple(w.numpy()))
                    elem.append(tf.convert_to_tensor(x, dtype='float32'))
        return tf.cast(tf.reshape(elem, self._output_shape), dtype='float32')

    if not training:
        self._data.append(inputs)
    return inputs

    # get values of the output of value map layer
def get_values(self):
    for d in self._data:
        try:
            d = d.numpy()
        except AttributeError:
            continue
        yield d

def get_config(self):
    config = super(valuemaplayer, self).get_config()
    config.update({"vmap": self._vmap.items(),
                   "compress": self._compression})
    return config

@classmethod
def from_config(cls, config):
    return cls(**config)

def compute_output_shape(self, input_shape):
    print("input shape of value map layer:", input_shape)
    self._output_shape=input_shape
    b, h, w, _ = input_shape
    if self._compression:
        print("channel axis is set")
        self._output_shape = tf.TensorShape((b, h, w, 1))
    return self._output_shape
#saving the model :
model.save('model.h5')
#load model :

new_model = tf.keras.models.load_model('model.h5',custom_objects = {"valuemaplayer": valuemaplayer})

**Comment by sushreebarsa** [comment]
@saylideshmukh Could you please provide the notebook link or colab gist to replicate the issue.I tried to replicate the issue and faced different error.Please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/1ad04357851608ff89922e34a1e5ac58/kerasgist16276.ipynb).
We see that you are using an older version of TF (2.3) which is not actively supported so we recommend you to kindly try with the latest TF versions and let us know the outcome?
Thank you!
